python/ray/serve/tests/unit/test_deployment_state.py

```diff
@@ -2701,5 +2701,440 @@ def test_get_active_node_ids_none(mock_deployment_state_manager_full):
+        deployment_state_update_result = deployment_state.update()
+        deployment_state._deployment_scheduler.schedule(
+            {deployment_state._id: deployment_state_update_result.upscale}, {}
+        )
+        check_counts(deployment_state, total=2, by_state=[(ReplicaState.STARTING, 2)])
+        assert deployment_state.curr_status_info.status == DeploymentStatus.UPDATING
+
+        for replica in deployment_state._replicas.get():
+            replica._actor.set_ready()
+
+        deployment_state.update()
+        check_counts(deployment_state, total=2, by_state=[(ReplicaState.RUNNING, 2)])
+        assert deployment_state.curr_status_info.status == DeploymentStatus.HEALTHY
+
+        deployment_state.update()
+        check_counts(deployment_state, total=2, by_state=[(ReplicaState.RUNNING, 2)])
+        assert deployment_state.curr_status_info.status == DeploymentStatus.HEALTHY
+
+        # Now update target_capacity to 100, should have no effect.
+        deployment_state.update(target_capacity=100)
+        check_counts(deployment_state, total=2, by_state=[(ReplicaState.RUNNING, 2)])
+        assert deployment_state.curr_status_info.status == DeploymentStatus.HEALTHY
+
+        # Now update target_capacity back to None, should have no effect.
+        deployment_state.update(target_capacity=None)
+        check_counts(deployment_state, total=2, by_state=[(ReplicaState.RUNNING, 2)])
+        assert deployment_state.curr_status_info.status == DeploymentStatus.HEALTHY
+
+    def test_target_capacity_0_1_replica(self, mock_deployment_state):
+        """
+        Deploy with target_capacity set to 0. Should have a single replica.
+        """
+        deployment_state, timer, cluster_node_info_cache = mock_deployment_state
+
+        b_info_1, b_version_1 = deployment_info(num_replicas=100)
+        updating = deployment_state.deploy(b_info_1)
+        assert updating
+
+        deployment_state_update_result = deployment_state.update(target_capacity=0)
+        deployment_state._deployment_scheduler.schedule(
+            {deployment_state._id: deployment_state_update_result.upscale}, {}
+        )
+        check_counts(deployment_state, total=1, by_state=[(ReplicaState.STARTING, 1)])
+        assert deployment_state.curr_status_info.status == DeploymentStatus.UPDATING
+
+        for replica in deployment_state._replicas.get():
+            replica._actor.set_ready()
+
+        deployment_state.update(target_capacity=0)
+        check_counts(deployment_state, total=1, by_state=[(ReplicaState.RUNNING, 1)])
+        assert deployment_state.curr_status_info.status == DeploymentStatus.HEALTHY
+
+        deployment_state.update(target_capacity=0)
+        check_counts(deployment_state, total=1, by_state=[(ReplicaState.RUNNING, 1)])
+        assert deployment_state.curr_status_info.status == DeploymentStatus.HEALTHY
+
+    def test_reduce_target_capacity(self, mock_deployment_state):
```

Comment:
What is an example of what target capacity would look like if a rollback happens? Can add a test for that?

Reply:
For now, we are not supporting in-place rollouts within the cluster, so it would just be reducing target_capacity for the target cluster and increasing it for the existing cluster (same path as the opposite direction)